### PR TITLE
Add camera support for linux (V4L2)

### PR DESCRIPTION
**Pull request by Florin9doi**

### CODE REVIEW

File: Core/HLE/sceKernel.cpp [hunk 1]
```diff
@@ -276,6 +276,7 @@ void __KernelDoState(PointerWrap &p)
 		__VideoPmpDoState(p);
 		__AACDoState(p);
 		__UsbGpsDoState(p);
+		// __UsbCamDoState(p); // not yet ;D
```
Comment (hrydgard) [inline Core/HLE/sceKernel.cpp]: Not yet, why? Either explain in the comment, or remove the line entirely please.

File: Core/HLE/sceUsbCam.cpp [hunk 2]
```diff
@@ -89,65 +126,43 @@ static int sceUsbCamReadMicBlocking(u32 bufAddr, u32 size) {
...
-	INFO_LOG(HLE, "UNIMPL sceUsbCamAutoImageReverseSW");
+static int sceUsbCamStopVideo() {
+	Camera::stopCapture();
 	return 0;
 }
 
 static int sceUsbCamReadVideoFrameBlocking(u32 bufAddr, u32 size) {
 	std::lock_guard<std::mutex> lock(videoBufferMutex);
-	u32 transferSize = std::min(videoBufferLength, size);
+	u32 transferSize = (std::min)(videoBufferLength, size);
```
Comment (hrydgard) [inline Core/HLE/sceUsbCam.cpp]: What's with the weird parentheses?

File: Core/HW/Camera.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+// https://github.com/hrydgard/ppsspp and http://www.ppsspp.org/.
+
+#pragma once
+
+#include "ppsspp_config.h"
+#include "Core/Config.h"
+#include "Log.h"
+
+#if PPSSPP_PLATFORM(LINUX) && !PPSSPP_PLATFORM(ANDROID)
+	#include <stdio.h>
```
Comment (hrydgard) [inline Core/HW/Camera.h]: nitpick: we don't indent includes

File: android/src/org/ppsspp/ppsspp/NativeActivity.java [hunk 4]
```diff
@@ -1409,12 +1409,6 @@ public boolean processCommand(String command, String params) {
 				// Only keep the screen bright ingame.
 				window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
 			}
-		} else if (command.equals("event")) {
```
Comment (hrydgard) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: Why is this removed?

**Review by hrydgard — CHANGES_REQUESTED** [review]

File: android/src/org/ppsspp/ppsspp/NativeActivity.java (hunk 4 above)
Comment (Florin9doi) [reply android/src/org/ppsspp/ppsspp/NativeActivity.java]: this is called from __UsbCamShutdown now

File: Core/HLE/sceKernel.cpp (hunk 1 above)
Comment (Florin9doi) [reply Core/HLE/sceKernel.cpp]: I find challenging to maintain backward compatibility for save states and I doubt that the current version contain enough information to cover every scenario. I'll remove it

File: Core/HW/Camera.h (hunk 3 above)
Comment (Florin9doi) [reply Core/HW/Camera.h]: fixed

File: Core/HLE/sceUsbCam.cpp (hunk 2 above)
Comment (Florin9doi) [reply Core/HLE/sceUsbCam.cpp]: https://stackoverflow.com/questions/5004858/why-is-stdmin-failing-when-windows-h-is-included . I replaced it with NOMINMAX

File: android/src/org/ppsspp/ppsspp/NativeActivity.java (hunk 4 above)
Comment (hrydgard) [reply android/src/org/ppsspp/ppsspp/NativeActivity.java]: No not this Android mCameraHelper, as far as I can tell skimming the code...

**Comment by hrydgard** [comment]
I'm fine with this now, except still unsure about that Android camerahelper thingy..

**Comment by Florin9doi** [comment]
event#exitgame was used to deallocate camera resources when a game is closed.
In this patch, the same thing is done by: 
__UsbCamShutdown>>Camera::stopCapture>>System_SendMessage("stopVideo")

**Comment by hrydgard** [comment]
I see! Sorry I missed that.

**Review by hrydgard — APPROVED** [review]
